rfc-022-makefile-standards.md

```diff
@@ -0,0 +1,81 @@
+# Makefile standards
+
+## Summary
+
+A set of standards for Makefiles across all projects, to supersede RFC012.
+
+## Problem
+
+When setting up/jumping onto a new project - sometimes it can be hard to discover what commands need to be run to perform common tasks (e.g. Running tests, serving the local development environment). This can make switching between projects more difficult as there can be a large number of ways to perform all of these tasks.
+
+Having a standard set of commands that can be run across all projects for common tasks allows people to get projects up and running without having to spend time looking into how things are set up to know what specific commands to run.
+
+## Proposal
+
+All projects **MUST** have a Makefile.
+
+All projects **MUST** have any relevant targets defined as [`PHONY` targets](https://www.gnu.org/software/make/manual/html_node/Phony-Targets.html) specified on just the first line of the file.
+
+All targets specified in this RFC **MUST** appear above any additional targets in the `Makefile`.
+
+**All** projects **MUST** have the commands:
+- `make test`
+  - Runs the tests for the project
+  - This **MAY** run the tests in a 'watch' mode (e.g. [Guard](https://github.com/guard/guard))
+  - This **MAY** have optional arguments for running a specific test/file
+- `make setup`
+  - Runs the setup for the project
+  - E.g. Runs `docker-compose build`
+- `make lint`
+  - Runs all the linting checks for the project
+  - E.g. Runs `rubocop`
+- `make lint-styling`
```

Comment:
curious as to when these extra lint commands come in useful?

Reply:
The Productionisation checklist has two separate scores as a measure of supportability for projects. If there are lots of linting offences, it's come in handy to tackle the style related lints first before the more complicated complexity related ones.

Reply:
I feel this mainly came about when dealing with existing projects that we were updating which had no linting run on, and shouldn't really be a thing in day-to-day development.

We could make it a bit more generic, however, and say related rules should be prefixed with these core names:

**ALL** projects **MUST** have the commands:
- `make test`
....

When a break down of these commands are needed, projects **SHOULD** use the related command as a prefix, e.g.:
```sh
make setup-db  # sets up the db
make setup-app  # sets up the app (deps, assets)
make lint-styling  # lints for styling issues
make lint-complexity  # lints for reducing complexity
make test-unit  # only runs unit tests
make test-watch  # runs tests as changes are made
make test-tag tag=feature
```